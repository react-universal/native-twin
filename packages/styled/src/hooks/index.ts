export { useComponentRegistration } from './useComponentRegistration';
export { useInteraction } from './useInteraction';
