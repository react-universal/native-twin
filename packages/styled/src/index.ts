export * from './styled';

export { useBuildStyledComponent } from './hooks/useStyledComponent';
export { useStyledContext } from './hooks/useStyledContext';

export * from './styled/variants';

export type {
  PropsFrom,
  StyledComponentProps,
  RegisteredComponent,
  ComponentStylesheet,
} from './types/styled.types';
export type { StyledContext, Units } from './types/css.types';

// CUSTOM
export { createStyledContext } from './utils/createStyledContext';
