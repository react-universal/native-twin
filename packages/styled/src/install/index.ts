export { setTailwindConfig } from '@universal-labs/stylesheets';
