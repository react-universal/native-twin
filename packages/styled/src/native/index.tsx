import React from 'react';
import * as RN from 'react-native';
import {
  styledFlatList,
  styledSectionList,
  styledVirtualizedList,
} from '../custom-components/List.components';
import createStyledComponent from './StyledComponent';

const styled = <StyleType, InitialProps extends { style?: RN.StyleProp<StyleType> }>(
  Component: React.ComponentType<InitialProps>,
) => createStyledComponent<StyleType, InitialProps>(Component);

styled.ActivityIndicator = styled(RN.ActivityIndicator);
styled.DrawerLayoutAndroid = styled(RN.DrawerLayoutAndroid);
styled.Image = styled(RN.Image);
styled.ImageBackground = styled(RN.ImageBackground);
styled.KeyboardAvoidingView = styled(RN.KeyboardAvoidingView);
styled.Modal = styled(RN.Modal);
styled.ScrollView = styled(RN.ScrollView);
styled.Switch = styled(RN.Switch);
styled.RefreshControl = styled(RN.RefreshControl);
styled.SafeAreaView = styled(RN.SafeAreaView);
styled.Text = styled(RN.Text);
styled.TextInput = styled(RN.TextInput);
styled.TouchableHighlight = styled(RN.TouchableHighlight);
styled.TouchableNativeFeedback = styled(RN.TouchableNativeFeedback);
styled.TouchableOpacity = styled(RN.TouchableOpacity);
styled.TouchableWithoutFeedback = styled(RN.TouchableWithoutFeedback);
styled.View = styled(RN.View);
styled.FlatList = styledFlatList;
styled.SectionList = styledSectionList;
styled.VirtualizedList = styledVirtualizedList;

export { styled as styledComponents };
