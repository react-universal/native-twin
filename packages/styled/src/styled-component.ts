import { forwardRef } from 'react';
import type { StyledComponentType } from './styled.types';
import useStyled from './useStyled';

export const StyledComponent = forwardRef(useStyled) as StyledComponentType;
