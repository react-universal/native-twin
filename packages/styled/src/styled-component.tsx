import { Gesture, GestureDetector } from 'react-native-gesture-handler';
import { ComponentType, forwardRef, ComponentProps } from 'react';
import type { IExtraProperties } from './styled.types';
import useStyled from './useStyled';

function styled<T extends ComponentType>(Component: T) {
  const Styled = forwardRef<T>(({ className = '', ...restProps }: any, ref) => {
    const { id, styles, gestures, hasInteractions, panHandlers } = useStyled(className);
    if (hasInteractions) {
      return (
        <GestureDetector gesture={Gesture.Exclusive(gestures.singleTap)}>
          <Component key={id} {...restProps} {...panHandlers} ref={ref} style={styles} />
        </GestureDetector>
      );
    }
    return <Component key={id} {...restProps} {...panHandlers} ref={ref} style={styles} />;
  });
  if (typeof Component !== 'string') {
    Styled.displayName = `StyledTW.${Component.displayName || Component.name || 'NoName'}`;
  }
  return Styled as ComponentType<ComponentProps<typeof Component> & IExtraProperties>;
}

export { styled };
export { mergeTWClasses } from './utils/mergeClasses';
