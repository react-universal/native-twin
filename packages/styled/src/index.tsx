import { ComponentType, forwardRef, useId, ComponentProps } from 'react';
import { useComponentRegistration, useInteraction } from './hooks';
import type { IExtraProperties } from './styled.types';

function styled<T extends ComponentType>(Component: T) {
  const Styled = forwardRef<T>(({ className = '', children, ...restProps }: any, ref) => {
    const componentID = useId();
    const { style } = useComponentRegistration(componentID, className, restProps?.style);
    const interaction = useInteraction(componentID, restProps);
    console.log('STYLE: ', style);
    return (
      <Component style={style} key={componentID} {...restProps} {...interaction} ref={ref}>
        {children}
      </Component>
    );
  });
  if (typeof Component !== 'string') {
    Styled.displayName = `StyledTW.${Component.displayName || Component.name || 'NoName'}`;
  }
  return Styled as ComponentType<ComponentProps<typeof Component> & IExtraProperties>;
}

export { styled };
export { mergeTWClasses } from './utils/mergeClasses';
