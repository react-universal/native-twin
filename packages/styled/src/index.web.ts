import styled from './styled';

export type {
  PropsFrom,
  StyledComponentProps,
  RegisteredComponent,
  ComponentStylesheet,
} from './types/styled.types';
export type { StyledContext, SheetMetadata, Units } from './types/css.types';
export default styled;
