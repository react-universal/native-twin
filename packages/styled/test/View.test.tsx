import renderer from 'react-test-renderer';
import { defineConfig, setup } from '@native-twin/native-twin';
import { presetTailwind } from '@native-twin/preset-tailwind';
import { View as StyledView, Text } from '../src';

beforeAll(() => {
  setup(defineConfig({ presets: [presetTailwind()] }));
});

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();
  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);
  return result as renderer.ReactTestRendererJSON;
}

describe('@native-twin/styled', () => {
  it('StyledView render', () => {
    const component = renderer.create(<StyledView className='flex-1' />);
    let tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});

describe('@native-twin/styled', () => {
  it('CustomView render', () => {
    const component = renderer.create(
      <StyledView className='shadow-sm web:p-10 sm:p-10 flex-1 outline-none'>
        <Text className='leading-6'>asd</Text>
      </StyledView>,
    );
    let tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
