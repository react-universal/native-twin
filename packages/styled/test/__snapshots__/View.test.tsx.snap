// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@native-twin/styled CustomView render 1`] = `
<View
  groupID=":r1:"
  style={
    {
      "elevation": 1,
      "flexBasis": "0%",
      "flexGrow": 1,
      "flexShrink": 1,
      "shadowColor": "rgb(0,0,0)",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.3,
      "shadowRadius": 3,
    }
  }
>
  <Text
    groupID=":r1:"
    parentID=":r1:"
    style={
      {
        "lineHeight": 1.5,
      }
    }
  >
    asd
  </Text>
</View>
`;

exports[`@native-twin/styled StyledView render 1`] = `
<View
  groupID=":r0:"
  style={
    {
      "flexBasis": "0%",
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
/>
`;
